[ao] Added function to inform dynamic vs static appropriate

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79326

Summary: The _detect_dynamic_vs_static function was added to take in a
prepared fx graph model that already had ModelReportObservers built into
it and uses the collected information to determine whether input and
output are stationary or non-stationary and provides feedback on whether
to make linear modules static or dynamic based on this information.

This PR will be followed up soon with another PR that will more
rigoursly test the whole end to end performance of this system, which is
primarily how the function in this PR will be tested for functionality,
which is why this one only has 1 test.

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao][work-in-progress] Added function to inform dynamic vs static appropriate

Summary: The _detect_dynamic_vs_static function was added to take in a
prepared fx graph model that already had ModelReportObservers built into
it and uses the collected information to determine whether input and
output are stationary or non-stationary and provides feedback on whether
to make linear modules static or dynamic based on this information.

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added function to inform dynamic vs static appropriate"


Summary: The _detect_dynamic_vs_static function was added to take in a
prepared fx graph model that already had ModelReportObservers built into
it and uses the collected information to determine whether input and
output are stationary or non-stationary and provides feedback on whether
to make linear modules static or dynamic based on this information.

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added function to inform dynamic vs static appropriate"


Summary: The _detect_dynamic_vs_static function was added to take in a
prepared fx graph model that already had ModelReportObservers built into
it and uses the collected information to determine whether input and
output are stationary or non-stationary and provides feedback on whether
to make linear modules static or dynamic based on this information.

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -726,3 +703,100 @@ def forward(self, x):
+"""
+
+
+class TestModelReportDetectDynamicStatic(QuantizationTestCase):
```
nit: please add `FX` to the name of the test case to indicate that this is related to fx

dzdang replied:
is this needed when this class is already in a file that has fx in the name in directory called fx? noticed we have a lot of code like that

vspenubarthi replied:
Decided to go ahead and make this change so it is consistent with the naming convention right now (and can be changed to adhere to any naming convention changes later)

jerryzh168 commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -134,9 +119,151 @@ def _detect_per_channel_helper(model: nn.Module):
+
+    Nonstationary distributions are below the tolerance level for this metric
+
+    This will then generate suggestions for dynamic vs static quantization focused around Linear
```
can we write a summary of the logic of how we recommend the quantization mode here?

jerryzh168 commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -134,9 +119,151 @@ def _detect_per_channel_helper(model: nn.Module):
+
+            if pre_stat > tolerance and post_stat > tolerance:
+                dynamic_recommended = False  # static is best if both stationary
+                pre_obs_dist_classification = DEFAULT_STATIONARY
+                post_obs_dist_classification = DEFAULT_STATIONARY
```
can we get these two just based on the relationship between pre_stat and tolerance/post_stat and tolerance? e.g.
pre_obs_dist_classification = ? if pre_stat > tolerance else ...
post_obs_dist_classification = ? if post_stat > tolerance else ...

jerryzh168 commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -134,9 +119,151 @@ def _detect_per_channel_helper(model: nn.Module):
+                pre_obs_dist_classification = DEFAULT_STATIONARY
+                post_obs_dist_classification = DEFAULT_NON_STATIONARY
+            else:
+                raise Exception("Should always take one of above branches")
```
we'll never reach here? if so we can change the last `elif` to else and comment with conditions

jerryzh168 commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -134,9 +119,151 @@ def _detect_per_channel_helper(model: nn.Module):
+                pre_obs_dist_classification = DEFAULT_STATIONARY
+                post_obs_dist_classification = DEFAULT_STATIONARY
+            elif pre_stat <= tolerance and post_stat > tolerance:
+                dynamic_recommended = False  # static best if input non-stationary, output stationary
```
we might be simplify the condition for dynamic_recommended as well I think

jerryzh168 commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -134,9 +119,151 @@ def _detect_per_channel_helper(model: nn.Module):
+            )
+
+        # if we have a non-stationary input -> linear -> stationary input we suggest
+        if (
```
can you add more details here? looks like we are making some recommendations here again, can we separate the code that we calculate the recommendation and the code that is used for printing the recommendation instead of mixing them?

vspenubarthi committed:
Update on "[ao] Added function to inform dynamic vs static appropriate"


Summary: The _detect_dynamic_vs_static function was added to take in a
prepared fx graph model that already had ModelReportObservers built into
it and uses the collected information to determine whether input and
output are stationary or non-stationary and provides feedback on whether
to make linear modules static or dynamic based on this information.

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added function to inform dynamic vs static appropriate"


Summary: The _detect_dynamic_vs_static function was added to take in a
prepared fx graph model that already had ModelReportObservers built into
it and uses the collected information to determine whether input and
output are stationary or non-stationary and provides feedback on whether
to make linear modules static or dynamic based on this information.

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -67,9 +63,7 @@ def _detect_per_channel_helper(model: nn.Module):
-                True
-                if sum(list(map(lambda x: isinstance(module, x), supported_modules))) > 0
-                else False
+                True if sum(list(map(lambda x: isinstance(module, x), supported_modules))) > 0 else False
```
Why not just
```
is_in_include_list = (
    sum(list(map(lambda x: isinstance(module, x), supported_modules))) > 0
)
```

andrewor14 commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -134,9 +119,136 @@ def _detect_per_channel_helper(model: nn.Module):
+
+    Args:
+        model (GraphModule): The prepared and calibrated GraphModule with inserted ModelReportObservers around layers of interest
+        tolerance (float, optional): The threshold where S metric is stationary above and non-stationary otherwise. Default: 0.5
```
Maybe document the return value as well?

andrewor14 commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -134,9 +119,136 @@ def _detect_per_channel_helper(model: nn.Module):
+DEFAULT_NON_STATIONARY = "non-stationary"
+
+
+def _detect_dynamic_vs_static(model: GraphModule, tolerance=0.5) -> Tuple[str, Dict[str, Any]]:
```
Is there any guidance on how to pick this value? I guess 0.5 is a reasonable one but I wonder if it's hard for the user to reason about what a good threshold is

vspenubarthi replied:
There really isn't for now, which is why we just decided to make it a parameter that the user could choose.

andrewor14 commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -134,9 +119,136 @@ def _detect_per_channel_helper(model: nn.Module):
+
+            # record module, pre and post stat, and whether to do dynamic or static based off it
+            # true if post observer data distribution is non-stationary, false if it's stationary
+            dynamic_recommended = True if post_stat <= tolerance else False
```
nit: just do
```
dynamic_recommended = post_stat <= tolerance
```

andrewor14 commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -134,9 +119,136 @@ def _detect_per_channel_helper(model: nn.Module):
+            quantization_type = "static"
+            quantization_reasoning = (
+                quantization_reasoning.format(
+                    module_name, module_info["pre_observer_data_dist"], module_info["post_observer_data_dist"]
```
These two lines are duplicated. Maybe something like this is cleaner?
```
if module_info["dynamic_recommended"]:
    quantization_type = "dynamic"
    benefit = dynamic_benefit
else:
    quantization_type = "static"
    benefit = static_benefit
quantitative_reasoning = quantitative_reasoning.format(...) + benefit
```

vspenubarthi committed:
Update on "[ao] Added function to inform dynamic vs static appropriate"


Summary: The _detect_dynamic_vs_static function was added to take in a
prepared fx graph model that already had ModelReportObservers built into
it and uses the collected information to determine whether input and
output are stationary or non-stationary and provides feedback on whether
to make linear modules static or dynamic based on this information.

This PR will be followed up soon with another PR that will more
rigoursly test the whole end to end performance of this system, which is
primarily how the function in this PR will be tested for functionality,
which is why this one only has 1 test.

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -134,9 +117,147 @@ def _detect_per_channel_helper(model: nn.Module):
+        tolerance (float, optional): The threshold where S metric is stationary above and non-stationary otherwise. Default: 0.5
+
+    Returns a tuple with two elements:
+        String report of of whether dynamic or static quantization is recommended for certain modules
```
I wonder if we should return a boolean here in case users want to do something programmatically based on this result (or build their own tools on top of this). Right now they would have to parse this string manually if they want to do that. What are your thoughts @HDCharles?

vspenubarthi replied:
There is a dictionary also returned that has the boolean value of whether it was recommended or not as well as other useful information such as the calculated metrics. I think this should contain the information you are discussing?

HDCharles commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -134,9 +117,147 @@ def _detect_per_channel_helper(model: nn.Module):
+    # loop through all submodules included nested ones
+    for name, module in model.named_modules():
+        # if module has the ModelReportObserver attached to it
+        if hasattr(module, DEFAULT_PRE_OBSERVER_NAME) and hasattr(module, DEFAULT_POST_OBSERVER_NAME):
```
this needs to check for supported modules, i.e. down the line where there are various reasons observers can be inserted, detecting pre/post may not be sufficient.

HDCharles commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -134,9 +117,147 @@ def _detect_per_channel_helper(model: nn.Module):
+                "post_observer_data_dist": post_obs_dist_classification,
+            }
+
+            module_dynamic_static_info[name] = module_info
```
name is not a unique identifier, need to use fqn

vspenubarthi replied:
this actually was the fqn (just printed and verified, renamed var to properly reflect)

HDCharles reviewed (approved):
see the additional comments, but otherwise looks good.

HDCharles commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -134,9 +117,147 @@ def _detect_per_channel_helper(model: nn.Module):
+    # store modules dynamic vs static information
+    module_dynamic_static_info = {}
+
+    # loop through all submodules included nested ones
```
would be good to put some higher level comments as well i.e. there are 2 big for loops in the code, the first I believe is looking for nodes and extracting all relevant information into module_dynamic_static_info and the second goes through module_dynamic_static_info and generates the recommendations.

vspenubarthi committed:
Update on "[ao] Added function to inform dynamic vs static appropriate"


Summary: The _detect_dynamic_vs_static function was added to take in a
prepared fx graph model that already had ModelReportObservers built into
it and uses the collected information to determine whether input and
output are stationary or non-stationary and provides feedback on whether
to make linear modules static or dynamic based on this information.

This PR will be followed up soon with another PR that will more
rigoursly test the whole end to end performance of this system, which is
primarily how the function in this PR will be tested for functionality,
which is why this one only has 1 test.

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportDetectDynamicStatic

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g
